server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java

```diff
@@ -58,53 +61,67 @@ public MappingMetadata(DocumentMapper docMapper) {
         this.routingRequired = docMapper.routingFieldMapper().required();
     }
 
-    @SuppressWarnings("unchecked")
-    public MappingMetadata(CompressedXContent mapping) {
-        this.source = mapping;
-        Map<String, Object> mappingMap = XContentHelper.convertToMap(mapping.compressedReference(), true).v2();
-        if (mappingMap.size() != 1) {
-            throw new IllegalStateException("Can't derive type from mapping, no root type: " + mapping.string());
+    // xcontent is always of type JSON and wrapped in a type object
+    public MappingMetadata(BytesReference mapping) {
```

Comment:
What if were to keep the `CompressedXContent` constructor instead here? I would feel more comfortable about that, as we currently implicitly assume that the bytes passed here are always decompressed?

Note that we are anyway decompressing the array again in the `new CompressedXContent()` call in line 83 as `CompressedXContent` is computing the `crc32`.